bundles/framework/featuredata/plugin/FeatureDataPluginHandler.js

```diff
@@ -0,0 +1,113 @@
+    getFeaturesByLayerId (layerId) {
+        const featuresMap = this.mapModule.getVectorFeatures(null, { layers: [layerId] });
+        return featuresMap && featuresMap[layerId] ? featuresMap[layerId].features : null;
+    }
+
+    createColumnSettingsFromFeatures (features) {
+        return Object.keys(features[0].properties)
+            .filter(key => !FEATUREDATA_DEFAULT_HIDDEN_FIELDS.includes(key))
+            .map(key => {
+                return {
+                    align: 'left',
+                    title: key,
+                    dataIndex: key,
+                    sorter: getSorterFor(key)
+                };
+            });
+    }
+
+    createDatasourceFromFeatures (features) {
+        return features.map(feature => {
+            return {
+                key: feature.properties.__fid,
+                ...feature.properties
+            };
+        });
+    }
+
+    createLayerTabs (layerId, layers, features) {
+        const tabs = layers?.map(layer => {
+            return {
+                key: layer.getId(),
+                label: layer.getName(),
+                children: layer.getId() === layerId && !!features && features.length
+                    ? createFeaturedataGrid(this.createColumnSettingsFromFeatures(features), this.createDatasourceFromFeatures(features))
+                    : null
+            };
+        }) || [];
+        return tabs;
+    }
+
+    setActiveTab (layerId) {
+        const featureDataLayers = this.getFeatureDataLayers() || null;
+        const features = layerId ? this.getFeaturesByLayerId(layerId) : null;
+        const tabs = this.createLayerTabs(layerId, featureDataLayers, features);
+        this.updateState({
+            activeTab: layerId,
+            tabs
+        });
+    }
+
+    updateStateAfterMapMoveEvent () {
+        const featureDataLayers = this.getFeatureDataLayers() || [];
+        const layerId = featureDataLayers?.find(layer => layer.getId() === this.state.activeTab) ? this.state.activeTab : featureDataLayers.map(layer => layer.getId())[0] || null;
+        const features = layerId ? this.getFeaturesByLayerId(layerId) : [];
+        const tabs = this.createLayerTabs(layerId, featureDataLayers, features);
+        this.updateState({
+            activeTab: layerId,
+            tabs
+        });
+    }
+
+    openFlyout () {
+        if (this.flyoutController) {
+            this.closeFlyout();
+            return;
+        }
+        const featureDataLayers = this.getFeatureDataLayers() || null;
+        const layerId = featureDataLayers && featureDataLayers.length ? featureDataLayers[0].getId() : null;
+        const features = layerId ? this.getFeaturesByLayerId(layerId) : null;
+        const tabs = this.createLayerTabs(layerId, featureDataLayers, features);
+        this.state = { tabs, controller: this.getController(), activeTab: layerId, onClose: () => this.getController().closeFlyout() };
+        this.flyoutController = showFeatureDataFlyout(this.state);
```

Comment:
Ei kannata suotta statea rakentaa tässä uudestaan ja laittaa sisään controlleria ja onClose callbackiä vaan annetaan ne vaan omina parametreinään/propseinaan:
```suggestion
        this.flyoutController = showFeatureDataFlyout(this.getState(), this.getController(),() => this.getController().closeFlyout());
```